code/messaging/src/phone/java/com/adobe/marketing/mobile/Messaging.java

```diff
@@ -191,20 +204,116 @@ public static void refreshInAppMessages() {
 
     // region proposition retrieval api
 
-    // TODO: implement
+    /**
+     * Registers a permanent event listener with the Mobile Core for listening to personalization decisions events received upon a personalization query to the Experience Edge network.
+     * If a new {@code AdobeCallback} is provided, it will replace the existing one and it will be invoked when propositions are received from the Edge network.
+     *
+     * @param callback A {@link AdobeCallback} which will be invoked with a {@link Map<Surface, List<Proposition>>} containing the {@link Surface}s and the corresponding list of {@link Proposition} objects.
+     */
     public static void setPropositionsHandler(@NonNull final AdobeCallback<Map<Surface, List<Proposition>>> callback) {
-
+        propositionsResponseHandler = callback;
+        if (!isPropositionsResponseListenerRegistered && callback != null) {
+            isPropositionsResponseListenerRegistered = true;
+            MobileCore.registerEventListener(EventType.MESSAGING, EVENT_SOURCE_NOTIFICATION, event -> {
+                final Map<Surface, List<Proposition>> convertedPropositions = new HashMap<>();
+                final Map<String, Object> eventData = event.getEventData();
+                if (!MapUtils.isNullOrEmpty(eventData)) {
+                    final List<Map<String, Object>> retrievedPropositionData = DataReader.optTypedListOfMap(Object.class, eventData, PROPOSITIONS, Collections.emptyList());
+                    for (final Map<String, Object> propositionData : retrievedPropositionData) {
+                        final Surface surface = Surface.fromUriString(DataReader.optString(propositionData, SCOPE, null));
+                        final List<Proposition> propositions = new ArrayList<>();
+                        final Proposition proposition = Proposition.fromEventData(propositionData);
+                        propositions.add(proposition);
+                        convertedPropositions.put(surface, propositions);
+                    }
+                }
+                propositionsResponseHandler.call(convertedPropositions);
```

Comment:
We are not invoking the call method here for empty propositions map.See https://github.com/adobe/aepsdk-optimize-android/blob/ajo-inbound-mobile/code/optimize/src/main/java/com/adobe/marketing/mobile/optimize/Optimize.java#L454